wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java

```diff
@@ -0,0 +1,1614 @@
+            }
+        }
+        return tags;
+    }
+
+    @Override
+    public void close() {
+        repository.close();
+    }
+
+    @Override
+    public File getWorkingDir() {
+        return repository.getWorkTree();
+    }
+
+    @Override
+    public List<RemoteReference> lsRemote(LsRemoteRequest request) throws UnauthorizedException, GitException {
+        String remoteUrl = request.getRemoteUrl();
+        LsRemoteCommand lsRemoteCommand = getGit().lsRemote();
+        lsRemoteCommand.setRemote(remoteUrl);
+        Collection<Ref> refs;
+        try {
+            refs = lsRemoteCommand.call();
+        } catch (GitAPIException exception) {
+            if (exception.getMessage().contains(ERROR_AUTHENTICATION_REQUIRED)) {
+                throw new UnauthorizedException(String.format(ERROR_AUTHENTICATION_FAILED, remoteUrl));
+            } else {
+                throw new GitException(exception.getMessage(), exception);
+            }
+        }
+        // Translate the JGit result
+        List<RemoteReference> remoteRefs = new ArrayList<>(refs.size());
+        for (Ref ref : refs) {
+            String commitId = ref.getObjectId().name();
+            String name = ref.getName();
+            RemoteReference remoteRef = newDto(RemoteReference.class).withCommitId(commitId).withReferenceName(name);
+            remoteRefs.add(remoteRef);
+        }
+        return remoteRefs;
+    }
+
+    @Override
+    public Config getConfig() throws GitException {
+        if (config != null) {
+            return config;
+        }
+        return config = new JGitConfigImpl(repository);
+    }
+
+    @Override
+    public void setOutputLineConsumerFactory(LineConsumerFactory outputPublisherFactory) {
+        //nothing to do, not outputs are produced by JGit
+    }
+
+    private Git getGit() {
+        if (git != null) {
+            return git;
+        }
+        return git = new Git(repository);
+    }
+
+    @Override
+    public List<String> listFiles(LsFilesRequest request) throws GitException {
+        return Arrays.asList(getWorkingDir().list(new FilenameFilter() {
+            @Override
+            public boolean accept(File dir, String name) {
+                return !name.startsWith(".");
+            }
+        }));
+    }
+
+    @Override
+    public void cloneWithSparseCheckout(String directory, String remoteUrl, String branch) throws GitException, UnauthorizedException {
+        //TODO rework this code when jgit will support sparse-checkout. Tracked issue: https://bugs.eclipse.org/bugs/show_bug.cgi?id=383772
+        clone(newDto(CloneRequest.class).withRemoteUri(remoteUrl));
+        if (!"master".equals(branch)) {
+            checkout(newDto(CheckoutRequest.class).withName(branch));
+        }
+        final String sourcePath = getWorkingDir().getPath();
+        final String keepDirectoryPath = sourcePath + "/" + directory;
+        IOFileFilter folderFilter = new DirectoryFileFilter() {
+      
```

Comment:
I don't know why but github move some of my comments to wrong line. So this comment I wrote for https://github.com/eclipse/che/pull/1297/files#diff-7f323c0167c366f41270f82a35ed368dR1586